### ISSUE TITLE
Factory provided DataAdapter.Fill throws InvalidOperationException

### ISSUE THREAD

**Issue by oising** [MiniProfiler/dotnet#313]
So, here's the pseudo-code (runtime is 4.7.2634.0):

```
var factory = DbProviderFactories.GetFactory("System.Data.SqlClient")
var profiledFactory = new ProfiledDbProviderFactory(factory);
var connection = profiledFactory.CreateConnection();
connection.ConnectionString = "..."

// ... other code here ... sqlCommand is a SqlCommand instance, with no connection info.
// it is create elsewhere - i.e. not created from profiledFactory 

var command = new ProfiledDbCommand(sqlCommand, connection, MiniProfiler.Current);
connection.Open()

var adapter = profiledFactory.CreateDataAdapter()
adapter.SelectCommand = command;
var data = new DataSet();

// this throws: IOE - ExecuteReader: Connection property has not been initialized.
adapter.Fill(data); 
```
When debugging, I can see that the connection is open. The ProfiledDbCommand's Connection property is using the ProfiledDbConnection instance. but the call stack blows up in SqlCommand.ValidateCommand(). This is because ProfileDbCommand does not assign the passed connection to the wrapped SqlCommand. 

This is a very old code base, hence the DataSet and weird patterns. And hence the profiling efforts :)

```
   at System.Data.SqlClient.SqlCommand.ValidateCommand(String method, Boolean async)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior)
   at StackExchange.Profiling.Data.ProfiledDbCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.FillInternal(DataSet dataset, DataTable[] datatables, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataSet dataSet, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataSet dataSet)
   at DataAccessClass.ExecuteSelect() in E:\oisin\Productivity.ERSUS\ERSUS\Classes\DataAccessClass.cs:line 199
   at GetRole.IsAdmin(String strUserPrincipalName) in E:\oisin\Productivity.ERSUS\ERSUS\Classes\GetRole.cs:line 24
   at ERSUS.Global.Session_Start(Object sender, EventArgs e) in E:\oisin\Productivity.ERSUS\ERSUS\Global.asax.cs:line 125
   at System.Web.SessionState.SessionStateModule.RaiseOnStart(EventArgs e)
   at System.Web.SessionState.SessionStateModule.CompleteAcquireState()
   at System.Web.SessionState.SessionStateModule.BeginAcquireState(Object source, EventArgs e, AsyncCallback cb, Object extraData)
   at System.Web.HttpApplication.AsyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
```

**Comment by oising** [comment]
Okay... so I dug deeper. And indeed, the `Connection` property of the wrapped `SqlCommand` is null . And the problem seems so simple as to be almost certainly an oversight on my behalf, but, please, indulge me. Looking at this constructor in `ProfiledDbCommand`: 

```
      public ProfiledDbCommand(DbCommand command, DbConnection connection, IDbProfiler profiler)
        {
            _command = command ?? throw new ArgumentNullException(nameof(command));
            _connection = connection;

            if (profiler != null)
            {
                _profiler = profiler;
            }
        }
```
This constructor should surely be setting the connection via the protected property `DbConnection` instead of setting the backing field `_connection` directly. Because when it does, it evades the logic in the property setter to correctly assign the connection to the wrapped command:

```
        protected override DbConnection DbConnection
        {
            get => _connection;
            set
            {
                _connection = value;
                if (value is ProfiledDbConnection profiledConn)
                {
                    _profiler = profiledConn.Profiler;
                    _command.Connection = profiledConn.WrappedConnection;
                }
                else
                {
                    _command.Connection = value;
                }
            }
        }
```

The IOE is thrown in ProfiledDbCommand.ExecuteDbReader( ... )

I'll put together a PR with a fix today.

**Comment by oising** [comment]
So, the key antagonist here is the fact that the SqlCommand was not created by the factory. Perhaps an edge case for code written from the ground up, but not so uncommon I'd say if someone's trying to shoehorn the profiler into an older codebase.

**Comment by oising** [comment]
I factored out the logic in the setter to avoid the virtual call in the constructor.

**Comment by oising** [comment]
Here's the test that encapsulates the fix:

```
        [Fact]
        public void DataReaderViaProfiledDbCommandWithNullConnection()
        {
            // https://github.com/MiniProfiler/dotnet/issues/313

            using (var conn = GetConnection())
            {
                var command = new SqliteCommand("select 1");
                var wrappedCommand = new ProfiledDbCommand(command, conn, conn.Profiler);

                var reader = wrappedCommand.ExecuteReader(CommandBehavior.SequentialAccess);
                Assert.True(reader.Read());
            }
        }
```

**Comment by NickCraver** [comment]
I need to dig on the problem here, I don't think it's what's proposed here exactly. The connection *is* set, because the underlying command comes from `DbConnection.CreateCommand()` ([happens here](https://github.com/MiniProfiler/dotnet/blob/7ebee3e8513e0110c95adade52fc2aab20bb06d0/src/MiniProfiler.Shared/Data/ProfiledDbConnection.cs#L145)). I think the problem is the code path you're calling probably doesn't work on anything but a `SqlConnection`, though it appears `DbConnection` works it really doesn't. I need to test this. A test showing it *breaking* would be the most useful PR at the moment, if possible to do. I need to understand what's broken before fixing it - will be back in action at a keyboard next week if all goes well.

**Comment by oising** [comment]
No, in my case, the command is NOT created via the profiled connection.
It's new'd up indépendantly in another library separate from yours and
outside of my control. Think of it like a weird CQRS pattern.

**Comment by oising** [comment]
I guarantee this is the issue and the fix. Because I have it in production
:)

**Comment by oising** [comment]
The test included in the PR is the case where it is breaking. Otherwise it
wouldnt be a very good test case, right? 😋

**Comment by NickCraver** [comment]
@oising There's a distinction between it fixes *this* issue, and it fixes things in general. For instance, `_connection` wasn't previously null after the constructor, but in your fix it now is. That'll break other use cases. As said before, I'll need to test this...and need my hands back in action to do it. I'm very reluctant to merge the current fix as-is.

**Comment by oising** [comment]
Oh lordy, I see what you're referring to. My bad. If course, the field _connection is left uninitialized. That's a mistake on my part. And I guess we need more tests because it's a hole if all of your tests passed nonetheless (which they did) 

I'll push another commit tomorrow. Well spotted.

**Comment by oising** [comment]
Well, after some brief analysis with Resharper, there are no usages of the protected property `ProfiledDbCommand.DbConnection` and none of its backing field either, so the underlying backing field `_connection` being `null` is not breaking anything _for now_. This is why it evades any test failures. Ultimately, you're quite correct: I unwittingly introduced a bug in waiting. I've pushed the remediation:

```
            if (connection != null)
            {
                _connection = connection;
                UnwrapAndAssignConnection(connection);
            }
```

**Comment by oising** [comment]
Bump. How are the wrists?

**Comment by NickCraver** [comment]
Still a bit slow here but getting back into it, thanks for asking :)

Fixed via #314 (thanks!), closing out to cleanup!
